### DISCUSSION TITLE
How do you add fixed point numbers in Zeptoforth

### DISCUSSION THREAD

**Discussion by rdoetjes** [tabemann/zeptoforth#54] [Q&A] [answered]
Hi there,

I need to add 0,5 before doing the (very convenient) 2 f.n
However in gforth I'd use f+ but that doesn't exist in zeptoforth (I grepped through the whole source). Now obviously the regular + word would not work in this instance.
So I am sure I am overlooking something here.

Thanks,

Ray

**Comment by tabemann** [comment, marked as answer]
To add and subtract fixed-point numbers you use the double-cell addition and subtraction words `d+` and `d-`, since there is no difference between fixed-point and double-cell addition and subtraction.

**Comment by rdoetjes** [comment]
Danke Schon! I didn't know that double-cell was identical to fp, nice and simple, I love it.

**Reply by tabemann** [reply]
Well, S31.32 fixed-point is equivalent to double-cell in this case ─ it is not equivalent to double-cell for multiplication or division, hence the existence of `f*` and `f/`.

**Comment by rdoetjes** [comment]
Yes I as using f/ and that's what puzzled me when I did a f+ that it wasn't there. And yes I felt kinda dumb not realising myself that d+ and d- work the same on double-cell :D
Thank you for your awesome support, and great documentation the wiki is probably the best off all the non-commercial embedded Forth systems out there. Much appreciated and so much fun touching Forth again after 30 years.

**Reply by tabemann** [reply]
Thank you for using zeptoforth! And there's no dumb questions ─ I am glad to support whoever uses zeptoforth however they need.

**Comment by robzed** [comment]
@rdoetjes I'm going to use these for clarity (and compatibility with a FlashForth project)

: fconstant [inlined] 2constant ; 
: f+ [inlined] d+ ;
: f- [inlined] d- ;
: fvariable [inlined] 2variable ;
: fdup [inlined] 2dup ;
: fswap [inlined] 2swap ;
: f! [inlined] 2! ;
: f@ [inlined] 2@ ;

**Reply by tabemann** [reply]
Note that there is no need to inline `fvariable`, since `2variable` does not normally get compiled into other words (even though in theory you can).

**Reply by rdoetjes** [reply]
That's funny robzed, I actually made a nested word for f+ and f- and fdup in my little project as well! 
I wasn't yet aware of the inline keyword, I will replace that a tiny bit more efficient. I would have used C's inline keyword in a similar situation, saves some branching. I am learning a lot here on Zeptoforth.
